docs/src/lib/functions.md

```diff
@@ -48,4 +48,62 @@ sort!
 tail
 unique!
 permutecols!
-```
\ No newline at end of file
+```
+
+## Indexing
+
+The following rules explain target functionality of how `getindex` and `view` are intended to work for `DataFrame`, `SubDataFrame` and `DataFrameRow`.
+
+The rules for a valid type of index into a column are the following:
+* a value, later denoted as `col`:
+    * a `Symbol`;
+    * an `Integer` that is not `Bool`;
+* a vector, later denoted as `cols`:
+    * a vector of `Symbol` (does not have to be a subtype of `AbstractVector{Symbol}`);
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`.
+    * a colon.
+
+The rules for a valid type of index into a row are the following:
+* a value, later denoted as `row`:
+    * an `Integer` that is not `Bool`;
+* a value, later denoted as `rows`:
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`;
+    * a colon.
+
+The following list specifies return types of `getindex` operations depending on argument types. In the summary `df` represents a `DataFrame`, `sdf` is a `SubDataFrame` and `dfr` is a `DataFrameRow`.
+
+`DataFrame`:
+* `df[col]` -> an alias of a vector contained in column `col`;
+* `df[cols]` -> a freshly allocated `DataFrame` containing aliases of vectors contained in columns `cols`;
+* `df[row, col]` -> a value contained in row `row` of column `col`;
+* `df[row, cols]` -> a `NamedTuple` containing data from row `row` in columns `cols`;
+* `df[rows, col]` -> a copy of a vector `df[col]` with only rows `rows` selected;
+* `df[rows, cols]` -> a `DataFrame` containing columns `cols` and `df[rows, col]` as a vector in each `col` in `cols`.
+* `@view df[col]` -> an alias of a vector contained in column `col`;
+* `@view df[cols]` -> a `SubDataFrame` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
```

Comment:
Maybe the special treatment of colon could be problematic, and we should also use `df[:]` as the parent in that case? Same in similar cases.

Reply:
Retaining `df` will be faster and I guess it will be important for `groupby` when you have a lot of small groups.

This is a case that currently was not on the table as `@view df[x]` treats `x` as rows, but it is going to be changed to ensure consistency.

As long as `df` is not mutated this should not matter in practice.

Reply:
"As long as `df` is not mutated" is what bothers me. :-)

Why do you expect `view(df, :)` to be used in `groupby`? Presumably if the overhead of calling `df[cols]` is OK when `cols != :`, it's also OK when `cols == :`? Or is the latter more common than the former?